### PR TITLE
parser: rewrite lex

### PR DESCRIPTION
**Pull request by tiancaiamao**
there is a basic new lexer now, See also https://github.com/pingcap/tidb/issues/1490
it can pass all tests in scanner_test.go and scan the most basic SQL statements.
but it still can't pass all tests in parser_test.go.

here is a benchmark run BenchmarkParse using the old lexer vs new lexer:
```
BenchmarkParse-4	   30000	     49745 ns/op	    5872 B/op	     117 allocs/op
BenchmarkParse-4	   30000	     44896 ns/op	    5672 B/op	     149 allocs/op
```

it's slightly faster,  that's a good beginning. 

the following work would be seperate test for lexer, make it (near) zero-alloc;
fullfill the lexer for more complex SQL;
finally, replace the old one until it's compatible enough.

### CODE REVIEW

File: parser/lexer.go [hunk 1]
```diff
@@ -0,0 +1,373 @@
+package parser
```
Comment (shenli) [inline parser/lexer.go]: Add license

File: parser/yy_parser.go [hunk 2]
```diff
@@ -66,12 +66,29 @@ type Parser struct {
...
 	cache     []yySymType
 	src       string
+	lexer     yyReset
+}
+
+type yyReset interface {
+	reset(sql string) yyLexerXX
+}
+
+type yyLexerXX interface {
```
Comment (shenli) [inline parser/yy_parser.go]: Weird name.

File: parser/lexer.go [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+
+	for _, v := range tokenPrefixTable {
+		if strings.HasPrefix(s.r.s[pos.Offset:], v.prefix) {
+			s.r.incN(len(v.prefix))
+			tok, lit = v.token, v.prefix
+			return
+		}
+	}
+
+	if ch0 == '@' {
```
Comment (coocood) [inline parser/lexer.go]: use switch is more readable.

File: parser/lexer.go [hunk 4]
```diff
@@ -0,0 +1,373 @@
...
+
+	if isIdentFirstChar(ch0) {
+		return s.scanIdent()
+	} else if isDigit(ch0) || ch0 == '.' {
+		return s.scanNumber()
+	} else if ch0 == '\'' || ch0 == '"' {
+		return s.scanString()
+	}
+
+	for _, v := range tokenPrefixTable {
```
Comment (coocood) [inline parser/lexer.go]: Loop through tokenPrefixTable every time is costly.

Comment (tiancaiamao) [inline parser/lexer.go]: Yes, I agree.
It's not best practice,  We could using a trie data struct to make a lookup table or using other method, but let's keep it simple currently. `tokenPrefixTable` is not big and I think we can bear it.

The new lexer's compatibility have higher priority, make it run first, then make it faster and make it elegant.

Comment (coocood) [inline parser/lexer.go]: Just a switch in a switch case will do, it's very simple too.

File: parser/yy_parser.go [hunk 5]
```diff
@@ -66,12 +66,24 @@ type Parser struct {
 	result    []ast.StmtNode
 	cache     []yySymType
 	src       string
+	lexer     yyReset
+}
+
+type yyReset interface {
```
Comment (shenli) [inline parser/yy_parser.go]: Please add comment for this interface.

**Comment by tiancaiamao** [comment]
after reduce object allocation and change longest prefix match to Trie, it's a bit faster then before:

```
BenchmarkParse-4	   30000	     49963 ns/op	    5888 B/op	     112 allocs/op
BenchmarkParse-4	   50000	     36257 ns/op	    5080 B/op	      85 allocs/op
```

**Comment by tiancaiamao** [comment]
now the new lexer can pass all tests in parser package!

**Comment by shenli** [comment]
Great! So can we drop the old parser? @tiancaiamao

**Comment by tiancaiamao** [comment]
now it pass all TiDB tests, and should be stable enough to handle 80% common cases.
I think we can merge it first, clean up code & fix bugs & more tests for compatible will be following PRs. 

PTAL @shenli @coocood @zimulala @hanfei19910905

File: parser/yy_parser.go [hunk 6]
```diff
@@ -68,15 +68,38 @@ type Parser struct {
 	result    []ast.StmtNode
 	cache     []yySymType
 	src       string
+	lexer     yyReset
+}
+
+// yyReset is an interface with reset method and returns a yyLexer.
```
Comment (shenli) [inline parser/yy_parser.go]: "lexerCreater" is a better name?

File: parser/parser_test.go [hunk 7]
```diff
@@ -465,7 +464,7 @@ func (s *testParserSuite) TestBuiltin(c *C) {
 		{"select sysdate(), sysdate(6)", true},
 		{"SELECT time('01:02:03');", true},
 
-		// Select current_time
+		// // Select current_time
```
Comment (zimulala) [inline parser/parser_test.go]: `// // ` -> `//`

File: parser/scanner_test.go [hunk 8]
```diff
@@ -442,3 +126,41 @@ SELECT`, selectKwd},
...
+			tok1 := l1.Lex(&v1)
+			tok2 := l2.Lex(&v2)
+			fmt.Println(tok1, tok2, v1, v2)
+			c.Assert(tok1, Equals, tok2)
+			if tok1 == 0 {
+				break
+			}
+		}
+	}
+
```
Comment (zimulala) [inline parser/scanner_test.go]: Remove this empty line.

File: parser/lexer.go [hunk 9]
```diff
@@ -0,0 +1,499 @@
...
+var _ = yyLexer(&Scanner{})
+
+// Pos represents the position of a token.
+type Pos struct {
+	Line   int
+	Col    int
+	Offset int
+}
+
+// Scanner implements the yyLexer interface
```
Comment (zimulala) [inline parser/lexer.go]: Add . at the end of a line

File: parser/yy_parser.go [hunk 10]
```diff
@@ -68,15 +68,38 @@ type Parser struct {
...
+// if an object implements both yyLexer and yyReset, it can return itself in the reset method, thus reducing allocation.
```
Comment (zimulala) [inline parser/yy_parser.go]: `if` -> `If`

**Comment by shenli** [comment]
Add a flag to switch new lexer on. So we can test it early.

File: parser/lexer.go [hunk 11]
```diff
@@ -0,0 +1,499 @@
...
+type Scanner struct {
+	r   reader
+	buf bytes.Buffer
+
+	errs         []error
+	stmtStartPos int
```
Comment (shenli) [inline parser/lexer.go]: How about subquery? or multiple statements in one sql text?

File: parser/lexer.go [hunk 12]
```diff
@@ -0,0 +1,499 @@
...
+	} else if ch0 == '`' {
+		return s.scanQuotedIdent()
+	}
+
+	switch ch0 {
+	case '@':
+		return s.startWithAt()
+	case '/':
+		return s.startWithSlash()
+	case '-':
```
Comment (shenli) [inline parser/lexer.go]: --i is an expression
-- i is an comment

Comment (shenli) [inline parser/lexer.go]: ignore this

File: parser/lexer.go [hunk 13]
```diff
@@ -0,0 +1,499 @@
...
+		}
+		ch0 := s.r.peek()
+		if ch0 == ending {
+			s.r.inc()
+			if s.r.peek() != ending {
+				break
+			}
+		}
+		// TODO this would break reader's line and col information
+		if ch0 == '\n' {
```
Comment (shenli) [inline parser/lexer.go]: Do nothing?

File: parser/lexer.go [hunk 14]
```diff
@@ -0,0 +1,499 @@
...
+		}
+		if ch0 == '\\' {
+			// save := s.r.pos()
```
Comment (shenli) [inline parser/lexer.go]: useless code?

File: parser/lexer.go [hunk 15]
```diff
@@ -0,0 +1,499 @@
...
+		} else if !isASCII(ch0) {
+			// TODO handle non-ascii
+		}
+		s.buf.WriteByte(ch0)
+		s.r.inc()
+	}
+
+	lit = s.buf.String()
+	// Quoted strings placed next to each other are concatenated to a single string.
+	// See http://dev.mysql.com/doc/refman/5.5/en/string-literals.html
```
Comment (shenli) [inline parser/lexer.go]: Use mysql 5.7 document reference.

File: parser/lexer.go [hunk 16]
```diff
@@ -0,0 +1,499 @@
...
+			} else if ch0 == '\\' {
+			} else if ch0 == '"' {
+				s.buf.WriteByte('"')
+				s.r.inc()
+				continue
+			} else if ch0 == 't' {
+				s.buf.WriteByte('\t')
+				s.r.inc()
+				continue
+			} else if ch0 == '\'' {
```
Comment (shenli) [inline parser/lexer.go]: For double quote, we should also consider '"'.

File: parser/lexer.go [hunk 17]
```diff
@@ -0,0 +1,499 @@
...
+	r.inc()
+	return
+}
+
+func (r *reader) pos() Pos {
+	return r.p
+}
+
+func (r *reader) data(from *Pos) string {
+	return r.s[from.Offset:r.p.Offset]
```
Comment (shenli) [inline parser/lexer.go]: Could from.Offset be greater than r.p.Offset?

File: parser/yy_parser.go [hunk 18]
```diff
@@ -124,16 +148,26 @@ func (parser *Parser) setLastSelectFieldText(st *ast.SelectStmt, lastEnd int) {
...
+
+	return v.offset
 }
 
-func (parser *Parser) endOffset(offset int) int {
-	offset--
+func (parser *Parser) endOffset(v *yySymType) int {
+	offset := v.offset
+	if _, ok := parser.lexer.(defaultLexer); ok {
+		// offset = v.offset + len(v.ident)
```
Comment (shenli) [inline parser/yy_parser.go]: useless code？

File: parser/yy_parser.go [hunk 19]
```diff
@@ -124,16 +148,26 @@ func (parser *Parser) setLastSelectFieldText(st *ast.SelectStmt, lastEnd int) {
 	}
 }
 
-func (parser *Parser) startOffset(offset int) int {
-	offset--
-	for unicode.IsSpace(rune(parser.src[offset])) {
-		offset++
+func (parser *Parser) startOffset(v *yySymType) int {
+	if _, ok := parser.lexer.(defaultLexer); ok {
+		offset := v.offset
```
Comment (shenli) [inline parser/yy_parser.go]: offset := v.offset - 1

File: parser/yy_parser.go [hunk 20]
```diff
@@ -92,9 +115,10 @@ func (parser *Parser) Parse(sql, charset, collation string) ([]ast.StmtNode, err
 	parser.result = parser.result[:0]
 
 	sql = handleMySQLSpecificCode(sql)
-	l := NewLexer(sql)
 
+	l := parser.lexer.reset(sql)
```
Comment (shenli) [inline parser/yy_parser.go]: parser.lexer.setSQL(sql)
yyParse(&parser.lexer, parser)

File: parser/lexer.go (hunk 13 above)
Comment (tiancaiamao) [inline parser/lexer.go]: nothing for the moment, add a TODO
this would change when I refact the code taking UTF8 into consideration

Comment (shenli) [inline parser/lexer.go]: So the if statement is useless.

File: parser/lexer.go (hunk 17 above)
Comment (tiancaiamao) [inline parser/lexer.go]: if that happens, it's a bug!
when you find any bugs, please report and I'll fix it.

**Comment by tiancaiamao** [comment]
scan() now get rid of switch case, completely... 
benchmark doesn't show any decrease compare to previous version

PTAL @coocood

File: parser/lexer.go (hunk 11 above)
Comment (tiancaiamao) [inline parser/lexer.go]: anything need special treat?
it can pass TestSubQuery test case.

**Comment by tiancaiamao** [comment]
fix issue https://github.com/pingcap/tidb/issues/1544
the new lexer will support UTF8 encoding...

`./bin/tidb-server -newlexer 1`

**Comment by shenli** [comment]
Great! @coocood PTAL

**Comment by coocood** [comment]
LGTM

File: parser/lexer.go (hunk 13 above)
Comment (shenli) [inline parser/lexer.go]: Should we remove this if? @tiancaiamao

File: parser/parser_test.go [hunk 21]
```diff
@@ -679,7 +677,7 @@ func (s *testParserSuite) TestDDL(c *C) {
 		{"CREATE TABLE foo (a bytes)", false},
 		{"CREATE TABLE foo (a SMALLINT UNSIGNED, b INT UNSIGNED)", true},
 		{"CREATE TABLE foo (a SMALLINT UNSIGNED, b INT UNSIGNED) -- foo", true},
-		{"CREATE TABLE foo (a SMALLINT UNSIGNED, b INT UNSIGNED) // foo", true},
+		// {"CREATE TABLE foo (a SMALLINT UNSIGNED, b INT UNSIGNED) // foo", true},
```
Comment (shenli) [inline parser/parser_test.go]: Why comment this line?

File: tidb-server/main.go [hunk 22]
```diff
@@ -36,15 +37,16 @@ import (
...
+	store       = flag.String("store", "goleveldb", "registered store name, [memory, goleveldb, boltdb, tikv]")
+	storePath   = flag.String("path", "/tmp/tidb", "tidb storage path")
+	logLevel    = flag.String("L", "info", "log level: info, debug, warn, error, fatal")
+	port        = flag.String("P", "4000", "mp server port")
+	statusPort  = flag.String("status", "10080", "tidb server status port")
+	lease       = flag.Int("lease", 1, "schema lease seconds, very dangerous to change only if you know what you do")
+	socket      = flag.String("socket", "", "The socket file to use for connection.")
+	enablePS    = flag.Int("perfschema", 0, "If enable performance schema.")
+	useNewPlan  = flag.Int("newplan", 1, "If use new planner.")
+	useNewLexer = flag.Int("newlexer", 0, "If use new lexer.")
```
Comment (shenli) [inline tidb-server/main.go]: use bool flag is better.

**Comment by shenli** [comment]
@tiancaiamao The branch has conflicts with master.

File: parser/parser_test.go (hunk 21 above)
Comment (tiancaiamao) [inline parser/parser_test.go]: this is a bad test case, `//` in SQL in not comment , while the old lexer treat it as comment.

File: parser/lexer.go (hunk 13 above)
Comment (tiancaiamao) [inline parser/lexer.go]: I'll do it next PR.
scanString should be improved: more clear, confirm to standard SQL stringLit, and avoid allocation if possible.

**Comment by shenli** [comment]
LGTM
Good job!

**Pull request merged**